### PR TITLE
✨ feat: Add generate quote function

### PR DESCRIPTION
**Pull request by HGZ-20**
- feat: Add generate quote function
- Adjusted the output of some error messages
- Use run_server() as a unified entry point for calling the enclave service
- Merged unnecessary commits

### CODE REVIEW

File: enclave/src/lib.rs [hunk 1]
```diff
@@ -210,27 +212,123 @@ fn update_rsa_key(path: &str) -> Result<(), sgx_status_t> {
...
+}
+
+fn init_rsa_key() -> Result<(), sgx_status_t> {
+    match read_rsa_key(RSA_KEY_PATH) {
+        Ok(_) => (),
+        Err(_) => {
+            println!("[+] Trying to generate a new RSA key.");
+            update_rsa_key(RSA_KEY_PATH)?;
+        }
+    };
```
Comment (jialez0) [inline enclave/src/lib.rs:228]: Following is better:
```rust
if read_rsa_key(RSA_KEY_PATH).is_err() {
    println!("[+] Trying to generate a new RSA key.");
    update_rsa_key(RSA_KEY_PATH)?;
}

File: enclave/src/lib.rs [hunk 2]
```diff
@@ -210,27 +212,123 @@ fn update_rsa_key(path: &str) -> Result<(), sgx_status_t> {
...
-#[no_mangle]
-pub unsafe extern "C" fn init_rsa_key() -> sgx_status_t {
-    let rsa_key_path = "RSA_KEY";
+fn generate_quote() -> Option<Vec<u8>>{
+    let mut rsa_key_data: RSAKeyData = (*RSA_KEY_DATA.lock().unwrap()).clone().unwrap();
 
-    match read_rsa_key(rsa_key_path) {
+    let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs();
+    // 5 minutes before the deadline still needs to update the key
+    if (rsa_key_data.not_befer < now as i64) | (rsa_key_data.not_after - ADVANCE_REFRESH_TIME < now as i64) {
```
Comment (jialez0) [inline enclave/src/lib.rs:239]: This line maybe too long, you can run `Cargo fmt` in the project directory to format your code.

File: enclave/src/lib.rs [hunk 3]
```diff
@@ -210,27 +212,123 @@ fn update_rsa_key(path: &str) -> Result<(), sgx_status_t> {
...
-        Err(_) => {
-            match update_rsa_key(rsa_key_path) {
-                Ok(_) => (),
-                Err(x) => return x
-            };
+        Err(x) => {
+            println!("[-] RSA key initialization failure.");
+            return x
+        }
+    };
```
Comment (jialez0) [inline enclave/src/lib.rs:320]: So as above, use `if init_rsa_key().is_err()` instead.

File: enclave/src/lib.rs [hunk 4]
```diff
@@ -210,27 +212,123 @@ fn update_rsa_key(path: &str) -> Result<(), sgx_status_t> {
...
+        sgx_status_t::SGX_SUCCESS => (),
+        _ => return None,
+    };
+
+    Some(return_quote_buf[0..quote_len as usize].to_vec())
+
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn run_server() -> sgx_status_t {
```
Comment (jialez0) [inline enclave/src/lib.rs:313]: There is no "server" now, so the better name of this function maybe `init()`

Comment (HGZ-20) [reply enclave/src/lib.rs:313]: I am now planning to write the rest of the service process in this ecall, so I have planned a run_server function in advance.

File: enclave/src/lib.rs (hunk 3 above)
Comment (HGZ-20) [reply enclave/src/lib.rs:320]: This contains a retrieval of the error content, so it doesn't seem appropriate to change it to if init_rsa_key().is_err() here.

**Comment by HGZ-20** [comment]
[Add logging](https://github.com/inclavare-containers/confidential-token-broker/pull/3/commits/24e49485fc8f32121c5c8a14e45c548aa7d62bd6)

# Key file not readable.
image

# Key file readable.
image

**Pull request merged**